nightfall-deployer/contracts/Shield.sol

```diff
@@ -18,225 +18,280 @@ import './Config.sol';
+        uint256 blockNumberL2,
+        Transaction[] memory ts,
+        uint256 index
+    ) external payable {
+        // The transaction is a withdrawal transaction
+        require(
+            ts[index].transactionType == TransactionTypes.WITHDRAW,
+            'Can only advance withdrawals'
         );
-    } else {
-      revert("Invalid Token Type");
+        // The block and transactions are real
+        state.isBlockReal(b, ts, blockNumberL2);
+
+        bytes32 withdrawTransactionHash = Utils.hashTransaction(ts[index]);
+        // The withdrawal has not been withdrawn
+        require(!withdrawn[withdrawTransactionHash], 'Cannot double withdraw');
+        address originalRecipientAddress = address(uint160(uint256(ts[index].recipientAddress)));
+        address currentOwner =
+            advancedWithdrawals[withdrawTransactionHash] == address(0)
+                ? originalRecipientAddress
+                : advancedWithdrawals[withdrawTransactionHash];
+
+        // Only the owner of the withdraw can set the advanced withdrawal
+        require(msg.sender == currentOwner, 'You are not the current owner of this withdrawal');
+        advancedFeeWithdrawals[withdrawTransactionHash] = msg.value;
+        payable(address(state)).transfer(msg.value);
+        emit InstantWithdrawalRequested(withdrawTransactionHash, msg.sender, msg.value);
+    }
+
+    function payOut(Transaction memory t, address recipientAddress) internal {
+        // Now pay out the value of the commitment
+        address addr = address(uint160(uint256(t.ercAddress)));
+        ERCInterface tokenContract = ERCInterface(addr);
+        // address recipientAddress = address(uint160(uint256(t.recipientAddress)));
+
+        if (t.tokenType == TokenType.ERC20) {
+            if (t.tokenId != ZERO) revert('ERC20 deposit should have tokenId equal to ZERO');
+            if (t.value >= super.getRestriction(addr))
+                revert('Value is above current restrictions for withdrawals');
+            else tokenContract.transfer(recipientAddress, uint256(t.value));
+        } else if (t.tokenType == TokenType.ERC721) {
+            if (t.value != 0)
+                // value should always be equal to 0
+                revert('Invalid inputs for ERC721 deposit');
+            else
+                tokenContract.safeTransferFrom(
+                    address(this),
+                    recipientAddress,
+                    uint256(t.tokenId),
+                    ''
+                );
+        } else if (t.tokenType == TokenType.ERC1155) {
+            tokenContract.safeTransferFrom(
+                address(this),
+                recipientAddress,
+                uint256(t.tokenId),
+                uint256(t.value),
+                ''
+            );
+        } else {
+            revert('Invalid Token Type');
+        }
+    }
+
+    function payIn(Transaction memory t) internal {
+        address addr = address(uint160(uint256(t.ercAddress)));
```

Comment:
This is a comment for future reference, no action required

The following code 
```
ERCInterface tokenContract = ERCInterface(
    address(uint160(uint256(t.ercAddress)))
);
```
has been replaced by this code
```
address addr = address(uint160(uint256(t.ercAddress)));
ERCInterface tokenContract = ERCInterface(addr);
```

Contracts have been refactored in a prior PR to make them gas efficient to reduce the cost of transaction for user.

This change shouldn't cause any gas changes because it is using the stack location in both scenarios. 
Similarly, in the future gas efficiency should be prioritised while making any changes to the contract.